### PR TITLE
Fixes bullseye conditions to be shown in SQ UI

### PR DESCRIPTION
**Pull request by tomassdepakin**
Hi. Fixing bug, when bullseye conditions are not shown in SonarQube UI. 
Tested with 6.3

Removing unneeded CoverageType from CoverageMeasure. It successfully contains line hits and conditions simultaneously.

Here comes my paint madskillz: 
bullseye_before

---
This change is [Reviewable](https://reviewable.io/reviews/sonaropencommunity/sonar-cxx/1237)

### CODE REVIEW

**Comment by Bertk** [comment]
👍 Thanks for this improvement. Did you also check it for SQ 5.6?
I could not find until now any reason why this should not work with SQ 5.6.
http://javadocs.sonarsource.org/5.6/apidocs/org/sonar/api/batch/sensor/coverage/NewCoverage.html

We missed this information (https://docs.sonarqube.org/display/DEV/API+Changes):

> **DEPRECATED** NewCoverage::ofType is deprecated and ignored. It is possible to feed as many coverage reports as you want for the same file. They will be merged without keeping any distinction of their "type".

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CoverageMeasures.java [hunk 1]
```diff
@@ -19,11 +19,7 @@
  */
 package org.sonar.cxx.sensors.coverage;
 
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Set;
-import java.util.Map.Entry;
+import java.util.*;
```
Comment (Bertk) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CoverageMeasures.java]: > squid:S2208 Wildcard imports should not be used

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CxxCoverageSensor.java [hunk 2]
```diff
@@ -334,23 +334,18 @@ private void saveMeasures(SensorContext context,
   /**
    * @param newCoverage
    * @param measure
-   * @param type
    */
-  private void checkCoverage(NewCoverage newCoverage, CoverageMeasure measure, CoverageMeasure.CoverageType type) {
+  private void checkCoverage(NewCoverage newCoverage, CoverageMeasure measure) {
     try {
-      if(type == CoverageMeasure.CoverageType.CONDITION) {
+        newCoverage.lineHits(measure.getLine(), measure.getHits());
```
Comment (Bertk) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CxxCoverageSensor.java]: > squid:IndentationCheck Make this line start at column 7

**Comment by Bertk** [comment]
The integration tests failed because the coverage values for SQ 5.6 changed (Travis and AppVeyor). 

```
Feature: Importing coverage data
  @SqApi56
  Scenario: Importing coverage reports
      Assertion Failed: 
        overall_line_coverage is actually 28.2 [expected: 20.0]
        line_coverage is actually 17.6 [expected: 12.5]
        it_line_coverage is actually 36.4 [expected: 26.3]
        coverage is actually 23.8 [expected: 20.0]
        overall_coverage is actually 34.0 [expected: 28.6]
        it_coverage is actually 40.6 [expected: 34.5]
        
  @SqApi56
  Scenario: Importing coverage reports zeroing coverage for untouched files 
      Assertion Failed: 
        overall_line_coverage is actually 20.0 [expected: 13.7]
        line_coverage is actually 7.3 [expected: 5.0]
        it_line_coverage is actually 19.0 [expected: 12.8]
        coverage is actually 11.1 [expected: 9.1]
        overall_coverage is actually 26.1 [expected: 21.5]
        it_coverage is actually 25.0 [expected: 20.4]
        

Feature: Smoketest
  @SqApi56
  Scenario: Smoketest
      Assertion Failed: 
        overall_coverage is actually 84.0 [expected: 81.8]
        it_coverage is actually 84.0 [expected: 81.8]
        coverage is actually 84.0 [expected: 81.8]

```

**Comment by guwirth** [comment]
@tomassdepakin thsnks for the fix.

In general looks good. Coverage is increased after your fix. It's always hard to say if coverage values are correct or not because every tool is doing/counting it al little bit different. Can you compare the numbers of our integration tests with the values shown in Bullseye please. Are they better/same/worse?

**Comment by guwirth** [comment]
@tomassdepakin files you have to adapt the values if they are right are:
* https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/integration-tests/features/importing_coverage.feature
* https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/integration-tests/features/smoketest.feature

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CoverageMeasures.java (hunk 1 above)
Comment (tomassdepakin) [reply cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CoverageMeasures.java]: Fixed

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CxxCoverageSensor.java (hunk 2 above)
Comment (tomassdepakin) [reply cxx-sensors/src/main/java/org/sonar/cxx/sensors/coverage/CxxCoverageSensor.java]: Fixed

**Comment by Bertk** [comment]
@tomassdepakin The integration tests still fail and you have to adapt the values within the scenarios for the Sonarqube 5.6 tests to pass the checks.

https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/integration-tests/features/importing_coverage.feature
https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/integration-tests/features/smoketest.feature
```
Feature: Importing coverage data
  @SqApi56
  Scenario: Importing coverage reports
      Assertion Failed: 
        overall_line_coverage is actually 28.2 [expected: 20.0]
        line_coverage is actually 17.6 [expected: 12.5]
        it_line_coverage is actually 36.4 [expected: 26.3]
        coverage is actually 23.8 [expected: 20.0]
        overall_coverage is actually 34.0 [expected: 28.6]
        it_coverage is actually 40.6 [expected: 34.5]
        
  @SqApi56
  Scenario: Importing coverage reports zeroing coverage for untouched files 
      Assertion Failed: 
        overall_line_coverage is actually 20.0 [expected: 13.7]
        line_coverage is actually 7.3 [expected: 5.0]
        it_line_coverage is actually 19.0 [expected: 12.8]
        coverage is actually 11.1 [expected: 9.1]
        overall_coverage is actually 26.1 [expected: 21.5]
        it_coverage is actually 25.0 [expected: 20.4]
        

Feature: Smoketest
  @SqApi56
  Scenario: Smoketest
      Assertion Failed: 
        overall_coverage is actually 84.0 [expected: 81.8]
        it_coverage is actually 84.0 [expected: 81.8]
        coverage is actually 84.0 [expected: 81.8]

```

**Comment by tomassdepakin** [comment]
Yep, I want to make more tests on coverage computing, and also with SQ5.6

**Comment by tomassdepakin** [comment]
Bullseye counts "for-loop", "try", "catch", "switch" as 1 decision, but we counted each of them as 2.
Fixed to treat such statements as 1 condition. 
Simplified condition\decision probes to be treated in same way. In the end SQ has only conditions.
Fixed uncovered functions to be laso shown in UI (previously weren't, see first post image L114).
After applying all fixes, uncovered conditions count in SQ UI is same as in bullseye reports.

Integration tests still to be fixed.

**Comment by tomassdepakin** [comment]
Checked with SQ 5.6 - UI and measures are same as in SQ 6.3
6.3:
6 3_measures_src_libraries_common_errors

5.6:
5 6_measures_src_libraries_common_errors

Overall coverage measures are very close to those in bullseye reports. In our project there is 1-5% difference per file.
Only difference left is line coverage. The more uncovered conditions - the more difference in overall coverage, because we mark them as "lineHits = 0" and SQ says to cover them not only as condition, but also as line.

**Comment by Bertk** [comment]
The coverage values for SQ6.2 and newer have now to be adopted as well.
```
  @SqApi62
  Scenario: Importing coverage reports
      	line_coverage is actually 25.0 [expected: 17.5]
      	coverage is actually 31.0 [expected: 25.9]
  @SqApi62
  Scenario: Smoketest 
      	coverage is actually 53.8 [expected: 50.0]
      	line_coverage is actually 54.8 [expected: 50.0]
```

**Comment by guwirth** [comment]
@tomassdepakin please rebase

**Comment by tomassdepakin** [comment]
@guwirth done

**Comment by guwirth** [comment]
@Bertk is final version also ok for you?

**Comment by guwirth** [comment]
@tomassdepakin looks good to me. Thanks for providing this PR.

**Comment by tomassdepakin** [comment]
Thank YOU guys for developing this plugin) My pleasure to participate.

**Pull request merged**